surfactant/filetypeid/id_extension.py

```diff
@@ -0,0 +1,38 @@
+# Copyright 2023 Lawrence Livermore National Security, LLC
+# See the top-level LICENSE file for details.
+#
+# SPDX-License-Identifier: MIT
+import pathlib
+import re
+from typing import Optional
+
+import surfactant.plugin
+
+
+@surfactant.plugin.hookimpl
+def identify_file_type(filepath: str) -> Optional[str]:
+    # pylint: disable=too-many-return-statements
+    try:
+        with open(filepath, "r") as f:
+            head = f.readline().strip("\n")
+            suffix = pathlib.Path(filepath).suffix.lower()
+            # Check for script files
+            if re.match(r"#!.*bash", head) or suffix == ".sh":
+                return "BASH"
+            if re.match(r"#!.*zsh", head) or suffix == ".zsh":
+                return "ZSH"
+            if re.match(r"#!.*php", head) or suffix == ".php":
+                return "PHP"
+            if re.match(r"#!.*python", head) or suffix == ".py":
```

Comment:
I saw [this post](https://stackoverflow.com/questions/2429511/why-do-people-write-usr-bin-env-python-on-the-first-line-of-a-python-script) and I think it might be a good idea to match just the script interpreter name in case the path is different across projects. Would this be fine or should it still include "/bin/" or "/usr/bin/env/" since it's very likely to follow this pattern anyways?

Reply:
Hm... okay, so I'm looking at things a bit more -- one common line is the shebang for running `/usr/bin/env bash` (note lack of a trailing slash after env, it's a program) for portability on different OSes.

What I'm thinking right now is, let's restructure this check slightly to split out the suffix checks into separate if statements, and then add a check for `head.startswith("#!")`, and then within that have a nested set of checks for what program is being run, with a default case that returns something like "SHEBANG".

Reply:
Another thing is for the `#!` check, `python3` should also be recognized as `PYTHON`.